### DISCUSSION TITLE
How to extract text with respect to the heading?

### DISCUSSION THREAD

**Discussion by Laxmi530** [jsvine/pdfplumber#797] [Ask for help with specific PDFs]
Hai,
Thank you for providing a beautiful library.
Actually, I am trying to extract portion of text with respect to the heading like in the sample pdf file we select the heading 'At glance 1' so as output need the text from 'The team' to ' penalty shootout.'. I tried some technique, but it is not working also tried to extract the headings so on the basis of position of headings we can process. 
so can someone please guide how to do this.

Thank you in advance.

[Sample_PDF_file.pdf](https://github.com/jsvine/pdfplumber/files/10510665/Sample_PDF_file.pdf)

**Comment by jsvine** [comment]
I would suggest an approach like this:

1. Use `page.extract_words(extra_attrs=["fontname", "size"])` to extract the words on the page, plus information about the character fonts and sizes. 
2. Use the information from step 1 to identify the positions of the headers.
3. Use the information from step 2 plus `page.crop(...).extract_text()` to get the text between each header. (The `...` will be a bounding box that you construct based on each consecutive pair of header positions.)

**Reply by Laxmi530** [reply]
@jsvine Thanks for the response.
With the help of `page.extract_words(extra_attrs=["fontname", "size"])` able to get the font size and font name but the problem is 
```
{'text': 'At', 'x0': 72.1, 'x1': 85.8903, 'top': 74.85124999999994, 'doctop': 74.85124999999994, 'bottom': 89.27625, 'upright': True, 'direction': 1, 'fontname': 'BCDEEE+Calibri-Bold', 'size': 14.425000000000068}
{'text': 'Glance', 'x0': 88.925, 'x1': 130.54850000000002, 'top': 74.85124999999994, 'doctop': 74.85124999999994, 'bottom': 89.27625, 'upright': True, 'direction': 1, 'fontname': 'BCDEEE+Calibri-Bold', 'size': 14.425000000000068}
{'text': 'Football,', 'x0': 72.1, 'x1': 110.8744, 'top': 100.5, 'doctop': 100.5, 'bottom': 111.69999999999993, 'upright': True, 'direction': 1, 'fontname': 'BCDFEE+Calibri', 'size': 11.199999999999932}
{'text': 'also', 'x0': 113.66319999999999, 'x1': 132.3448, 'top': 100.5, 'doctop': 100.5, 'bottom': 111.69999999999993, 'upright': True, 'direction': 1, 'fontname': 'BCDFEE+Calibri', 'size': 11.199999999999932}
{'text': 'called', 'x0': 134.43919999999997, 'x1': 161.07279999999997, 'top': 100.5, 'doctop': 100.5, 'bottom': 111.69999999999993, 'upright': True, 'direction': 1, 'fontname': 'BCDFEE+Calibri', 'size': 11.199999999999932}
{'text': 'association', 'x0': 162.4168, 'x1': 213.83599999999998, 'top': 100.5, 'doctop': 100.5, 'bottom': 111.69999999999993, 'upright': True, 'direction': 1, 'fontname': 'BCDFEE+Calibri', 'size': 11.199999999999932}
{'text': 'football', 'x0': 215.1576, 'x1': 249.70959999999997, 'top': 100.5, 'doctop': 100.5, 'bottom': 111.69999999999993, 'upright': True, 'direction': 1, 'fontname': 'BCDFEE+Calibri', 'size': 11.199999999999932}
{'text': 'or', 'x0': 251.92719999999997, 'x1': 261.436, 'top': 100.5, 'doctop': 100.5, 'bottom': 111.69999999999993, 'upright': True, 'direction': 1, 'fontname': 'BCDFEE+Calibri', 'size': 11.199999999999932}
```
you can see `At Glance` this is one heading the output i got that is in two dictionary so combining these two again a challenge so is there any method so that i can get font size and font name as per line.

**Reply by jsvine** [reply]
For situations like this, you can take advantage of the fact that `doctop` will be the same value for text of the same size on the same line.

**Reply by danltw** [reply]
@jsvine wondering if there is documentation on how to use `doctop` as mentioned above. I'm doing something similar where I need to group texts with respect to the heading.

@Laxmi530 are you able to share your solution to me as a reference?

**Reply by jsvine** [reply]
@danltw The solution I suggested is something you'd just handle through standard Python programming, rather than being specific to `pdfplumber`. (E.g., you could use `groupby(words, key="doctop")`, or a similar paradigm.) In the time since, however, I've added `Page.extract_text_lines(...)`, which might suit your needs.

**Reply by danltw** [reply]
Thanks @jsvine , managed to do it with "doctop"

**Reply by Laxmi530** [reply]
Hai @danltw,
Sorry for the late response.
What I did actually 1st count the font name and font size 2nd drop the most repeated font name and font size and consider the remaining font name and font size on that basics you will get the heading names. Post getting the heading names you can apply indexing. This is not a more generic way to extract the text with respect to the heading.

Can you please share your approach?

Thank you.

**Reply by danltw** [reply]
@Laxmi530 My code is an adaptation from this article https://towardsdatascience.com/extracting-headers-and-paragraphs-from-pdf-using-pymupdf-676e8421c467

However, I couldn't continue with PyMuPDF as it wasn't permissive in licensing, thus changed to pdfplumber instead.

The tl;dr approach is as follows (as adapted from above article):
- First iteration through the doc to extract the `fontname` and `size` attributes, doing a count of the unique combinations of `fontname_fontsize`.
- Then do a reverse sort so that the highest used combination of `fontname_fontsize` aka first element in the sorted list will be taken as the default paragraph font size. With this information, I can safely tag this combination as a paragraph, in my case, I chose to prepend html tags - `<p>` paragraphs, `<h1>` headers, `<s2>` subscripts. This is just my way of tagging, you may choose something else that makes sense and can be easily identified.
- Current iteration of my code will do 2 steps... first would be assuming that headers fonts would be larger than the presumed paragraph font, and subscripts smaller. This is a simple comparator of `fontsize` from the `fontname_fontsize` identifiers from above.
- However, I found that my sample PDFs have headers that are the same size as the paragraph fontsize. Hence, I currently assume bold as headers for now, and do a cleanup when I am joining back the words into lines downstream

Sorry in advance, kinda bad in explaining

**Reply by Laxmi530** [reply]
Thanks @danltw you explained well.
will try your solution.

Thanks for your explanation.

**Reply by jaideep11061982** [reply]
@danltw  would you like to provide the link for your approach ?

**Comment by bansalsahab** [comment]
> glanc

hey,
did you get the code?

**Comment by lpi-tn** [comment]
I recommend you to use pdfplumber + https://pypi.org/project/refinedoc/

You just need to format data from pdfplumber for the lib and it's work
